pkg/instancetype/upgrade.go

```diff
@@ -0,0 +1,159 @@
+//nolint:lll
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+package instancetype
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	instancetypeapi "kubevirt.io/api/instancetype"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+)
+
+type Upgrader interface {
+	Upgrade(*appsv1.ControllerRevision) (*appsv1.ControllerRevision, error)
+}
+
+var _ Upgrader = &upgrader{}
+
+type upgrader struct {
+	client     kubecli.KubevirtClient
+	vmInformer cache.SharedIndexInformer
+}
+
+func NewUpgrader(client kubecli.KubevirtClient, vmInformer cache.SharedIndexInformer) *upgrader {
+	return &upgrader{
+		client:     client,
+		vmInformer: vmInformer,
+	}
+}
+
+func (u *upgrader) Upgrade(original *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
+	// If the CR already contains an object from the latest version of the
+	// instancetype.kubevirt.io API *and* has the label showing this then skip
+	if IsObjectLatestVersion(original) {
+		return original, nil
+	}
+
+	// Find the VM owner of the CR
+	vm, err := u.discoverOwner(original)
+	if err != nil {
+		return nil, err
+	}
+
+	// Upgrade the CR object and generate a new CR
+	newCR, err := u.upgradeControllerRevision(vm, original)
+	if err != nil {
+		return nil, err
+	}
+
+	newCR, err = u.client.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), newCR, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Update owner reference and delete original CR if CR names do not equal
+	if newCR.Name != original.Name {
```

Comment:
Can this condition still be `false`? If so the `Create` before would have failed already.

Reply:
I don't believe so unless someone manually removes the labels from a `ControllerRevision` created after this PR lands and we then try to upgrade the `ControllerRevision` while it contains an object of the latest version.

I don't think we need to defend against that situation here as it complicates the code, the upgrade would just fail in that case and could be easily fixed.